File: packages/java/src/main/java/io/opensms/ApiTransport.java
package io.opensms;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.type.TypeFactory;
import io.opensms.models.Page;

import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;

/**
 * The single place that talks to the network. Owns authentication and standard
 * headers, JSON encoding and decoding, the {@code Idempotency-Key} header, retries
 * with backoff (honouring {@code Retry-After}), and mapping non-2xx responses to
 * {@link OpensmsException}. Resources are thin and delegate here. Not part of the
 * public API.
 */
final class ApiTransport {

    static final String USER_AGENT = "opensms-java/" + OpensmsClient.VERSION;

    private static final Set<Integer> RETRY_STATUSES = Set.of(429, 500, 502, 503, 504);
    private static final long MAX_RETRY_AFTER_SECONDS = 60;
    private static final long BACKOFF_BASE_MS = 500;
    private static final long BACKOFF_CAP_MS = 8000;

    private final String apiKey;
    private final String baseUrl;
    private final int maxRetries;
    private final HttpTransport http;
    private final Sleeper sleeper;

    ApiTransport(String apiKey, String baseUrl, int maxRetries, HttpTransport http, Sleeper sleeper) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
        this.maxRetries = maxRetries;
        this.http = http;
        this.sleeper = sleeper;
    }

    // -- type helpers -------------------------------------------------------

    static JavaType type(Class<?> c) {
        return TypeFactory.defaultInstance().constructType(c);
    }

    static JavaType listOf(Class<?> element) {
        return Json.MAPPER.getTypeFactory().constructCollectionType(List.class, element);
    }

    static JavaType pageOf(Class<?> element) {
        return Json.MAPPER.getTypeFactory().constructParametricType(Page.class, element);
    }

    static final JavaType VOID = type(Void.class);

    /** The explicit key from {@code opts}, else a fresh UUIDv4 (generated once per call). */
    static String idempotencyKey(RequestOptions opts) {
        if (opts != null && opts.getIdempotencyKey() != null) {
            return opts.getIdempotencyKey();
        }
        return UUID.randomUUID().toString();
    }

    // -- entry points used by resources ---------------------------------------

    <T> T get(String path, Query query, JavaType type) {
        return execute("GET", path + (query == null ? "" : query.build()), null, null, null, type);
    }

    /**
     * JSON request. A POST without {@code idempotencyKey} is never retried.
     *
     * @param idempotencyKey header value, or {@code null} to send none.
     */
    <T> T json(String method, String path, Object body, String idempotencyKey, JavaType type) {
        byte[] bytes = null;
        if (body != null) {
            try {
                bytes = Json.MAPPER.writeValueAsBytes(body);
            } catch (IOException e) {
                throw new IllegalArgumentException("request body could not be encoded as JSON", e);
            }
        }
        return execute(method, path, bytes, bytes == null ? null : "application/json", idempotencyKey, type);
    }

    <T> T raw(String method, String path, byte[] body, String contentType, String idempotencyKey, JavaType type) {
        return execute(method, path, body, contentType, idempotencyKey, type);
    }

    // -- core -------------------------------------------------------------------

    private <T> T execute(String method, String pathAndQuery, byte[] body, String contentType,
                          String idempotencyKey, JavaType type) {
        Map<String, String> headers = new LinkedHashMap<>();
        headers.put("Authorization", "Bearer " + apiKey);
        headers.put("Accept", "application/json");
        headers.put("User-Agent", USER_AGENT);
        if (contentType != null) {
            headers.put("Content-Type", contentType);
        }
        if (idempotencyKey != null) {
            headers.put("Idempotency-Key", idempotencyKey);
        }
        HttpTransport.Request request = new HttpTransport.Request(
                method, URI.create(baseUrl + pathAndQuery), headers, body);
        boolean safeToRepeat = !"POST".equals(method) || idempotencyKey != null;

        for (int attempt = 1; ; attempt++) {
            boolean mayRetry = safeToRepeat && attempt <= maxRetries;
            HttpTransport.Response res;
            try {
                res = http.send(request);
            } catch (IOException e) {
                if (mayRetry) {
                    pause(backoff(attempt));
                    continue;
                }
                throw OpensmsException.local("OpenSMS request failed: " + e, null, e);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw OpensmsException.local("OpenSMS request interrupted", null, e);
            }

            int status = res.status();
            if (status >= 200 && status < 300) {
                return decode(res, type);
            }
            OpensmsException err = Errors.fromResponse(res);
            if (mayRetry && RETRY_STATUSES.contains(status)) {
                Long retryAfter = err.getRetryAfter();
                if (retryAfter != null) {
                    if (retryAfter > MAX_RETRY_AFTER_SECONDS) {
                        throw err;
                    }
                    pause(Duration.ofSeconds(retryAfter));
                } else {
                    pause(backoff(attempt));
                }
                continue;
            }
            throw err;
        }
    }

    /** Full-jitter exponential backoff: {@code random(0, min(8 s, 0.5 s * 2^(n-1)))}. */
    static Duration backoff(int retryNumber) {
        long ceiling = Math.min(BACKOFF_CAP_MS, BACKOFF_BASE_MS << Math.min(retryNumber - 1, 20));
        return Duration.ofMillis(ThreadLocalRandom.current().nextLong(ceiling + 1));
    }

    private void pause(Duration d) {
        try {
            sleeper.sleep(d);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw OpensmsException.local("OpenSMS request interrupted", null, e);
        }
    }

    @SuppressWarnings("unchecked")
    private static <T> T decode(HttpTransport.Response res, JavaType type) {
        if (type.getRawClass() == Void.class || res.status() == 204 || res.body().length == 0) {
            return null;
        }
        try {
            return (T) Json.MAPPER.readValue(res.body(), type);
        } catch (IOException e) {
            String text = new String(res.body(), StandardCharsets.UTF_8);
            throw new OpensmsException("OpenSMS response could not be decoded: " + e.getMessage(),
                    res.status(), null, null, null, null, null, null, null, null, text, e);
        }
    }
}

File: packages/dotnet/src/Internal/ApiTransport.cs
using System;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;

namespace Opensms.Internal
{
    /// <summary>How a request uses the Idempotency-Key header.</summary>
    internal enum Idempotency
    {
        /// <summary>Never send the header.</summary>
        None,
        /// <summary>Always send one: the caller's key or a generated UUIDv4.</summary>
        Send,
    }

    /// <summary>One logical API call (possibly several HTTP attempts).</summary>
    internal sealed class ApiRequest
    {
        public HttpMethod Method { get; }
        public string Path { get; }
        public string Query { get; set; } = "";
        public object? JsonBody { get; set; }
        public Func<HttpContent>? Content { get; set; }
        public Idempotency Idempotency { get; set; } = Idempotency.None;
        public string? IdempotencyKey { get; set; }

        public ApiRequest(HttpMethod method, string path)
        {
            Method = method;
            Path = path;
        }
    }

    /// <summary>
    /// The only code that touches the network. Owns authentication headers,
    /// JSON encoding, Idempotency-Key handling, retries with backoff, per-attempt
    /// timeouts and mapping non-2xx responses to <see cref="OpensmsException"/>.
    /// </summary>
    internal sealed class ApiTransport : IDisposable
    {
        public const string DefaultBaseUrl = "https://opensms.io";
        public const string Version = "0.1.1";
        public const string UserAgent = "opensms-dotnet/" + Version;

        private static readonly HttpMethod Patch = new HttpMethod("PATCH");
        private static readonly TimeSpan MaxRetryAfter = TimeSpan.FromSeconds(60);

        private readonly string _apiKey;
        private readonly HttpClient _http;
        private readonly bool _ownsHttp;
        private readonly Func<TimeSpan, CancellationToken, Task> _sleep;

        public string BaseUrl { get; }
        public int MaxRetries { get; }
        public TimeSpan Timeout { get; }

        public ApiTransport(string apiKey, OpensmsClientOptions options)
        {
            _apiKey = apiKey;
            BaseUrl = (options.BaseUrl ?? DefaultBaseUrl).TrimEnd('/');
            if (!Uri.TryCreate(BaseUrl, UriKind.Absolute, out _))
                throw new ArgumentException($"baseUrl '{options.BaseUrl}' is not an absolute URL", nameof(options));
            MaxRetries = options.MaxRetries < 0 ? throw new ArgumentOutOfRangeException(nameof(options), "maxRetries must be >= 0") : options.MaxRetries;
            Timeout = options.Timeout <= TimeSpan.Zero ? throw new ArgumentOutOfRangeException(nameof(options), "timeout must be positive") : options.Timeout;
            _sleep = options.Sleep ?? ((d, ct) => Task.Delay(d, ct));

            if (options.HttpClient != null)
            {
                _http = options.HttpClient;
                _ownsHttp = false;
            }
            else
            {
                // Per-attempt timeouts are enforced with a CancellationToken below.
                _http = options.HttpMessageHandler != null
                    ? new HttpClient(options.HttpMessageHandler, disposeHandler: false)
                    : new HttpClient();
                _http.Timeout = System.Threading.Timeout.InfiniteTimeSpan;
                _ownsHttp = true;
            }
        }

        public static HttpMethod PatchMethod => Patch;

        /// <summary>Send and decode a JSON response body. Returns <c>default</c> for an empty body or 204.</summary>
        public async Task<T> SendAsync<T>(ApiRequest request, CancellationToken ct)
        {
            var text = await SendRawAsync(request, ct).ConfigureAwait(false);
            if (string.IsNullOrWhiteSpace(text)) return default!;
            try
            {
                return Json.Deserialize<T>(text!)!;
            }
            catch (System.Text.Json.JsonException ex)
            {
                throw new OpensmsException(0, $"OpenSMS response could not be decoded: {ex.Message}", body: text, innerException: ex);
            }
        }

        /// <summary>Send a request whose success response has no useful body (204).</summary>
        public Task SendNoContentAsync(ApiRequest request, CancellationToken ct) => SendRawAsync(request, ct);

        private async Task<string?> SendRawAsync(ApiRequest request, CancellationToken ct)
        {
            // Generated once per call and reused on every retry: that is what makes POST retries safe.
            string? idempotencyKey = request.Idempotency == Idempotency.Send
                ? (string.IsNullOrEmpty(request.IdempotencyKey) ? Guid.NewGuid().ToString() : request.IdempotencyKey)
                : null;
            var retryable = request.Method != HttpMethod.Post || idempotencyKey != null;
            var jsonBytes = request.JsonBody != null ? Json.Serialize(request.JsonBody) : null;
            var url = BaseUrl + request.Path + request.Query;

            for (var attempt = 0; ; attempt++)
            {
                ct.ThrowIfCancellationRequested();
                var canRetry = retryable && attempt < MaxRetries;
                using var message = new HttpRequestMessage(request.Method, url);
                message.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
                message.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
                message.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
                if (idempotencyKey != null) message.Headers.TryAddWithoutValidation("Idempotency-Key", idempotencyKey);
                if (jsonBytes != null)
                {
                    var content = new ByteArrayContent(jsonBytes);
                    content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
                    message.Content = content;
                }
                else if (request.Content != null)
                {
                    message.Content = request.Content();
                }

                using var attemptCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
                attemptCts.CancelAfter(Timeout);

                HttpResponseMessage response;
                string text;
                try
                {
                    response = await _http.SendAsync(message, HttpCompletionOption.ResponseContentRead, attemptCts.Token).ConfigureAwait(false);
                    text = await response.Content.ReadAsStringAsync(attemptCts.Token).ConfigureAwait(false);
                }
                catch (OperationCanceledException) when (ct.IsCancellationRequested)
                {
                    throw; // the caller cancelled: never retry or wrap
                }
                catch (Exception ex) when (ex is HttpRequestException || ex is OperationCanceledException || ex is IOException)
                {
                    if (canRetry)
                    {
                        await _sleep(Backoff(attempt + 1), ct).ConfigureAwait(false);
                        continue;
                    }
                    var reason = ex is OperationCanceledException ? $"timed out after {Timeout.TotalSeconds:0.#} s" : ex.Message;
                    throw new OpensmsException(0, $"OpenSMS request failed: {reason}", innerException: ex);
                }

                using (response)
                {
                    var status = (int)response.StatusCode;
                    if (status >= 200 && status < 300)
                        return status == 204 ? null : text;

                    var retryAfter = ParseRetryAfter(response);
                    if (canRetry && IsRetryStatus(status) && (retryAfter == null || retryAfter.Value <= MaxRetryAfter))
                    {
                        await _sleep(retryAfter ?? Backoff(attempt + 1), ct).ConfigureAwait(false);
                        continue;
                    }

                    throw OpensmsException.FromResponse(
                        status,
                        text,
                        Header(response, "X-Request-ID"),
                        retryAfter?.TotalSeconds);
                }
            }
        }

        private static bool IsRetryStatus(int status)
            => status == 429 || status == 500 || status == 502 || status == 503 || status == 504;

        /// <summary>Full-jitter exponential backoff: random(0, min(8 s, 0.5 s * 2^(n-1))).</summary>
        internal static TimeSpan Backoff(int retryNumber)
        {
            var cap = Math.Min(8.0, 0.5 * Math.Pow(2, retryNumber - 1));
            return TimeSpan.FromSeconds(Random.Shared.NextDouble() * cap);
        }

        private static TimeSpan? ParseRetryAfter(HttpResponseMessage response)
        {
            var typed = response.Headers.RetryAfter;
            if (typed?.Delta != null) return typed.Delta;
            if (typed?.Date != null)
            {
                var d = typed.Date.Value - DateTimeOffset.UtcNow;
                return d < TimeSpan.Zero ? TimeSpan.Zero : d;
            }
            var raw = Header(response, "Retry-After");
            if (raw != null && double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture, out var secs) && secs >= 0)
                return TimeSpan.FromSeconds(secs);
            return null;
        }

        private static string? Header(HttpResponseMessage response, string name)
            => response.Headers.TryGetValues(name, out var values) ? values.FirstOrDefault() : null;

        public void Dispose()
        {
            if (_ownsHttp) _http.Dispose();
        }
    }
}
